projects/lesson-5/runtime/src/kitties.rs

```diff
@@ -0,0 +1,435 @@
+}
+
+#[cfg_attr(feature = "std", derive(Debug, PartialEq, Eq))]
+#[derive(Encode, Decode)]
+pub struct KittyLinkedItem<T: Trait> {
+	pub prev: Option<T::KittyIndex>,
+	pub next: Option<T::KittyIndex>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Kitties {
+		/// Stores all the kitties, key is the kitty id / index
+		pub Kitties get(kitty): map T::KittyIndex => Option<Kitty<T>>;
+		/// Stores the total number of kitties. i.e. the next kitty index
+		pub KittiesCount get(kitties_count): T::KittyIndex;
+
+		pub OwnedKitties get(owned_kitties): map (T::AccountId, Option<T::KittyIndex>) => Option<KittyLinkedItem<T>>;
+	  KittyOwner get(owner_of): map T::KittyIndex => Option<T::AccountId>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Create a new kitty
+		pub fn create(origin) {
+			let sender = ensure_signed(origin)?;
+			let kitty_id = Self::next_kitty_id()?;
+
+			// Generate a random 128bit value
+			let dna = Self::random_value(&sender);
+
+			// Create and store kitty
+			let kitty =  Kitty {
+                dna: dna,
+                price: 0.into(),
+            };
+			Self::insert_kitty(&sender, kitty_id, kitty);
+		}
+
+		/// Breed kitties
+		pub fn breed(origin, kitty_id_1: T::KittyIndex, kitty_id_2: T::KittyIndex) {
+			let sender = ensure_signed(origin)?;
+
+			Self::do_breed(&sender, kitty_id_1, kitty_id_2)?;
+		}
+
+		// 作业：实现 transfer(origin, to: T::AccountId, kitty_id: T::KittyIndex)
+		// 使用 ensure! 来保证只有主人才有权限调用 transfer
+		// 使用 OwnedKitties::append 和 OwnedKitties::remove 来修改小猫的主人
+		pub fn transfer(origin, to: T::AccountId, kitty_id: T::KittyIndex) {
+			let sender = ensure_signed(origin)?;
+
+			let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+        ensure!(owner == sender, "You do not own this kitty");			
+
+			Self::transfer_from(sender, to, kitty_id);
+		}
+
+		fn set_price(origin, kitty_id: T::KittyIndex, new_price: T::Balance) -> Result {
+					let sender = ensure_signed(origin)?;
+
+					ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+					let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+					ensure!(owner == sender, "You do not own this cat");
+
+					let mut kitty = Self::kitty(kitty_id).unwrap();
+					kitty.price = new_price;
+
+					<Kitties<T>>::insert(kitty_id, kitty);
+
+					Ok(())
+    }
+
+		fn buy_kitty(origin, kitty_id: T::KittyIndex, max_price: T::Balance) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner != sender, "You can't buy your own cat");
+
+            let mut kitty = Self::kitty(kitty_id).unwrap();
+
+            let kitty_price = kitty.price;
+            ensure!(!kitty_price.is_zero(), "The cat you want to buy is not for sale");
```

Comment:
Using zero price is ok for now, but a better way is we have a flag or storage to identify which kitty is for sale.